src/wast-parser.cc

```diff
@@ -1686,10 +1686,39 @@ Result WastParser::ParsePlainInstr(std::unique_ptr<Expr>* out_expr) {
     case TokenType::SimdShuffleOp: {
       Token token = Consume();
       ErrorUnlessOpcodeEnabled(token);
-      Const const_;
-      CHECK_RESULT((ParseSimdConst(&const_, Type::I32, sizeof(v128))));
+      v128 value;
+      char* value_ptr = reinterpret_cast<char*>(&value);
```

Comment:
Use the `Bitcast` function, and a `uint8_t[16]` destination instead.

Reply:
`Bitcast<uint8_t[16]>` is not supported.

Reply:
So I've used a `uint8_t[16]` to store the values, and `Bitcast<v128>` to bitcast to it from the `v128`, but see below, the other way around does not appear to be supported.

Reply:
Oh right, probably because of pointer decay.

Reply:
So.. kind of. Template argument deduction will deduce their type correctly as arrays, but as you mention, they decay into pointers, so overload resolution picks overloads taking pointers as well. 

What truly generic code does is add a couple of specific overloads to match array types before pointer decay: 

* range-v3 https://github.com/ericniebler/range-v3/blob/master/include/range/v3/begin_end.hpp#L305
* libc++ https://github.com/llvm-mirror/libcxx/blob/master/include/iterator#L1667

But treating C arrays properly kind of requires unit-testing all APIs against them, and unless one is writing a C++ standard library, this is probably not worth doing.